// js-draw/dist/cjs/commands/lib.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.uniteCommands = exports.invertCommand = exports.SerializableCommand = exports.Erase = exports.Duplicate = exports.Command = void 0;
const Command_1 = __importDefault(require("./Command"));
exports.Command = Command_1.default;
const Duplicate_1 = __importDefault(require("./Duplicate"));
exports.Duplicate = Duplicate_1.default;
const Erase_1 = __importDefault(require("./Erase"));
exports.Erase = Erase_1.default;
const invertCommand_1 = __importDefault(require("./invertCommand"));
exports.invertCommand = invertCommand_1.default;
const SerializableCommand_1 = __importDefault(require("./SerializableCommand"));
exports.SerializableCommand = SerializableCommand_1.default;
const uniteCommands_1 = __importDefault(require("./uniteCommands"));
exports.uniteCommands = uniteCommands_1.default;
