// js-draw/dist/cjs/tools/localization.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultToolLocalization = void 0;
exports.defaultToolLocalization = {
    penTool: (penId) => `Pen ${penId}`,
    selectionTool: 'Selection',
    selectAllTool: 'Select all shortcut',
    eraserTool: 'Eraser',
    touchPanTool: 'Touch panning',
    twoFingerPanZoomTool: 'Panning and zooming',
    undoRedoTool: 'Undo/Redo',
    rightClickDragPanTool: 'Right-click drag',
    pipetteTool: 'Pick color from screen',
    keyboardPanZoom: 'Keyboard pan/zoom shortcuts',
    selectionMenu__show: 'Show selection menu',
    selectionMenu__copyToClipboard: 'Copy to clipboard',
    selectionMenu__duplicate: 'Duplicate',
    selectionMenu__delete: 'Delete',
    selectionMenu__paste: 'Paste',
    copyPasteError__heading: 'Copy/paste',
    copyPasteError__description: 'Something went wrong — this tool may not have clipboard access.',
    copyPasteError__errorDetails: 'Show error',
    copyPasteError__pasteRetry: 'To retry, please paste into the input box below:',
    copyPasteError__copyRetry: 'To retry, please copy the text in the input box below:',
    copyPasteError__copyMe: 'Copy me!',
    autocorrectedTo: (strokeDescription) => `Autocorrected to ${strokeDescription}`,
    autocorrectionCanceled: 'Autocorrect cancelled',
    textTool: 'Text',
    enterTextToInsert: 'Text to insert',
    changeTool: 'Change tool',
    pasteHandler: 'Copy paste handler',
    soundExplorer: 'Sound-based image exploration',
    disableAccessibilityExploreTool: 'Disable sound-based exploration',
    enableAccessibilityExploreTool: 'Enable sound-based exploration',
    soundExplorerUsageAnnouncement: 'Sound-based image exploration enabled: Click/drag the screen to play a sound representation of different parts of the image.',
    findLabel: 'Find',
    toNextMatch: 'Next',
    closeDialog: 'Close',
    findDialogShown: 'Find dialog shown',
    findDialogHidden: 'Find dialog hidden',
    focusedFoundText: (matchIdx, totalMatches) => `Viewing match ${matchIdx} of ${totalMatches}`,
    anyDevicePanning: 'Any device panning',
    copied: (count) => `Copied ${count} item(s)`,
    pasted: (count) => `Pasted ${count} item(s)`,
    toolEnabledAnnouncement: (toolName) => `${toolName} enabled`,
    toolDisabledAnnouncement: (toolName) => `${toolName} disabled`,
};
